internal/pkg/agent/application/upgrade/artifact/download/http/verify_backoff_rtt_test.go

```diff
@@ -0,0 +1,83 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package http
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+func TestVerifyBackoffRoundtripper(t *testing.T) {
+	t.Run("test get request retry", func(t *testing.T) {
+		failedResCounter := 2
+		handler := func(rw http.ResponseWriter, req *http.Request) {
+			if failedResCounter > 0 {
+				rw.WriteHeader(http.StatusInternalServerError)
+				failedResCounter--
+			}
+			_, err := rw.Write([]byte("hello"))
+			require.NoError(t, err)
+		}
+		server := httptest.NewServer(http.HandlerFunc(handler))
+		client := http.Client{
+			Transport: WithBackoff(&http.Transport{}, logp.NewLogger("testing")),
+			Timeout:   10 * time.Second,
+		}
+
+		res, err := client.Get(server.URL) //nolint:noctx // test code
+		require.NoError(t, err)
+		defer res.Body.Close()
+
+		body, err := io.ReadAll(res.Body)
+		require.NoError(t, err)
+
+		require.Equal(t, string(body), "hello")
+		require.Equal(t, res.StatusCode, 200)
+		require.Equal(t, failedResCounter, 0)
+	})
+
+	t.Run("test post request with body", func(t *testing.T) {
+		failedResCounter := 2
+		handler := func(rw http.ResponseWriter, req *http.Request) {
+			if failedResCounter > 0 {
+				rw.WriteHeader(http.StatusInternalServerError)
+				failedResCounter--
+			}
+
+			body, err := io.ReadAll(req.Body)
```

Comment:
[Suggestion]
Here and after L76, you could validate that the request/response body contain the correct value. So if there is a bug when reading the body and replacing it by the `bytes.Reader` we can catch it.

Reply:
On line 79 I'm validating that I receive the same value that is sent to the test server. So I am already doing what you're suggesting here. 

I'm not sure how necessary it is to validate the body in line 57 as it is only a test function. Please let me know if I'm misunderstanding anything here.